File: backend/src/controllers/employee.controller.js
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();

class EmployeeController {
  // Get all employees
  async getAllEmployees(req, res) {
    try {
      const { page = 1, limit = 10, status, departmentId, search } = req.query;
      
      const where = {};
      if (status) where.status = status;
      if (departmentId) where.departmentId = parseInt(departmentId);
      if (search) {
        where.OR = [
          { name: { contains: search, mode: 'insensitive' } },
          { employeeCode: { contains: search, mode: 'insensitive' } },
          { workEmail: { contains: search, mode: 'insensitive' } }
        ];
      }
      
      const [employees, total] = await Promise.all([
        prisma.employee.findMany({
          where,
          include: {
            department: true,
            jobPosition: true,
            manager: { select: { id: true, name: true, employeeCode: true } }
          },
          skip: (page - 1) * limit,
          take: parseInt(limit),
          orderBy: { createdAt: 'desc' }
        }),
        prisma.employee.count({ where })
      ]);
      
      res.json({
        success: true,
        data: employees,
        pagination: {
          total,
          page: parseInt(page),
          limit: parseInt(limit),
          totalPages: Math.ceil(total / limit)
        }
      });
    } catch (error) {
      res.status(500).json({
        success: false,
        message: 'Error fetching employees',
        error: error.message
      });
    }
  }
  
  // Get employee by ID
  async getEmployeeById(req, res) {
    try {
      const { id } = req.params;
      
      const employee = await prisma.employee.findUnique({
        where: { id: parseInt(id) },
        include: {
          user: { select: { email: true, role: true, isActive: true } },
          department: true,
          jobPosition: true,
          manager: { select: { id: true, name: true, employeeCode: true } },
          subordinates: { select: { id: true, name: true, employeeCode: true } },
          attendance: { take: 10, orderBy: { date: 'desc' } },
          leaves: { take: 5, orderBy: { createdAt: 'desc' } }
        }
      });
      
      if (!employee) {
        return res.status(404).json({
          success: false,
          message: 'Employee not found'
        });
      }
      
      res.json({
        success: true,
        data: employee
      });
    } catch (error) {
      res.status(500).json({
        success: false,
        message: 'Error fetching employee',
        error: error.message
      });
    }
  }
  
  // Create employee
  async createEmployee(req, res) {
    try {
      const employeeData = req.body;
      
      // Generate employee code
      const lastEmployee = await prisma.employee.findFirst({
        orderBy: { id: 'desc' }
      });
      const nextNumber = lastEmployee ? parseInt(lastEmployee.employeeCode.replace('EMP', '')) + 1 : 1;
      const employeeCode = `EMP${String(nextNumber).padStart(5, '0')}`;
      
      const employee = await prisma.employee.create({
        data: {
          ...employeeData,
          employeeCode,
          dateOfJoining: new Date(employeeData.dateOfJoining)
        },
        include: {
          department: true,
          jobPosition: true
        }
      });
      
      res.status(201).json({
        success: true,
        message: 'Employee created successfully',
        data: employee
      });
    } catch (error) {
      res.status(500).json({
        success: false,
        message: 'Error creating employee',
        error: error.message
      });
    }
  }
  
  // Update employee
  async updateEmployee(req, res) {
    try {
      const { id } = req.params;
      const updateData = req.body;
      
      const employee = await prisma.employee.update({
        where: { id: parseInt(id) },
        data: updateData,
        include: {
          department: true,
          jobPosition: true
        }
      });
      
      res.json({
        success: true,
        message: 'Employee updated successfully',
        data: employee
      });
    } catch (error) {
      res.status(500).json({
        success: false,
        message: 'Error updating employee',
        error: error.message
      });
    }
  }
  
  // Delete employee
  async deleteEmployee(req, res) {
    try {
      const { id } = req.params;
      
      await prisma.employee.delete({
        where: { id: parseInt(id) }
      });
      
      res.json({
        success: true,
        message: 'Employee deleted successfully'
      });
    } catch (error) {
      res.status(500).json({
        success: false,
        message: 'Error deleting employee',
        error: error.message
      });
    }
  }
  
  // Get employee statistics
  async getStatistics(req, res) {
    try {
      const [total, active, inactive, onLeave, byDepartment] = await Promise.all([
        prisma.employee.count(),
        prisma.employee.count({ where: { status: 'ACTIVE' } }),
        prisma.employee.count({ where: { status: 'INACTIVE' } }),
        prisma.employee.count({ where: { status: 'ON_LEAVE' } }),
        prisma.employee.groupBy({
          by: ['departmentId'],
          _count: true
        })
      ]);
      
      res.json({
        success: true,
        data: {
          total,
          active,
          inactive,
          onLeave,
          byDepartment
        }
      });
    } catch (error) {
      res.status(500).json({
        success: false,
        message: 'Error fetching statistics',
        error: error.message
      });
    }
  }
}

module.exports = new EmployeeController();

File: backend/addons/hr_employee_management/controllers/employee_controller.py
# -*- coding: utf-8 -*-
from odoo import http
from odoo.http import request
import json

class EmployeeController(http.Controller):
    """API Controller for Employee Management"""
    
    @http.route('/api/employee/list', type='json', auth='user', methods=['POST'])
    def get_employee_list(self, **kwargs):
        """Get list of all employees"""
        try:
            domain = kwargs.get('domain', [])
            fields = kwargs.get('fields', ['name', 'employee_code', 'work_email', 'department_id'])
            limit = kwargs.get('limit', 100)
            
            employees = request.env['hr.employee'].search_read(
                domain=domain,
                fields=fields,
                limit=limit
            )
            
            return {
                'success': True,
                'data': employees,
                'count': len(employees)
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    @http.route('/api/employee/create', type='json', auth='user', methods=['POST'])
    def create_employee(self, **kwargs):
        """Create new employee"""
        try:
            data = kwargs.get('data', {})
            employee = request.env['hr.employee'].create(data)
            
            return {
                'success': True,
                'message': 'Employee created successfully',
                'employee_id': employee.id,
                'employee_code': employee.employee_code
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    @http.route('/api/employee/<int:employee_id>', type='json', auth='user', methods=['POST'])
    def get_employee_details(self, employee_id, **kwargs):
        """Get employee details by ID"""
        try:
            employee = request.env['hr.employee'].browse(employee_id)
            
            if not employee.exists():
                return {
                    'success': False,
                    'error': 'Employee not found'
                }
            
            return {
                'success': True,
                'data': {
                    'id': employee.id,
                    'name': employee.name,
                    'employee_code': employee.employee_code,
                    'email': employee.work_email,
                    'mobile': employee.work_phone,
                    'department': employee.department_id.name if employee.department_id else None,
                    'job_title': employee.job_id.name if employee.job_id else None,
                    'status': employee.active_status,
                }
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    @http.route('/api/employee/update', type='json', auth='user', methods=['POST'])
    def update_employee(self, **kwargs):
        """Update employee information"""
        try:
            employee_id = kwargs.get('employee_id')
            data = kwargs.get('data', {})
            
            employee = request.env['hr.employee'].browse(employee_id)
            
            if not employee.exists():
                return {
                    'success': False,
                    'error': 'Employee not found'
                }
            
            employee.write(data)
            
            return {
                'success': True,
                'message': 'Employee updated successfully'
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    @http.route('/api/employee/delete', type='json', auth='user', methods=['POST'])
    def delete_employee(self, **kwargs):
        """Delete employee"""
        try:
            employee_id = kwargs.get('employee_id')
            employee = request.env['hr.employee'].browse(employee_id)
            
            if not employee.exists():
                return {
                    'success': False,
                    'error': 'Employee not found'
                }
            
            employee.unlink()
            
            return {
                'success': True,
                'message': 'Employee deleted successfully'
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    @http.route('/api/employee/search', type='json', auth='user', methods=['POST'])
    def search_employee(self, **kwargs):
        """Search employees by name, code, or email"""
        try:
            search_term = kwargs.get('search_term', '')
            
            domain = ['|', '|',
                     ('name', 'ilike', search_term),
                     ('employee_code', 'ilike', search_term),
                     ('work_email', 'ilike', search_term)]
            
            employees = request.env['hr.employee'].search_read(
                domain=domain,
                fields=['name', 'employee_code', 'work_email', 'department_id'],
                limit=50
            )
            
            return {
                'success': True,
                'data': employees,
                'count': len(employees)
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
